### ISSUE TITLE
Derive Foldable, Functor, and Traversable instances

### ISSUE THREAD

**Issue by aymannadeem** [tree-sitter/haskell-tree-sitter#182]
This was originally a task we conceived of when adding annotations, but it's a bit orthogonal and deserves its own special limelight. 

#### Problem

When I tried to automatically derive these instances ie., `deriveClause = [ DerivClause Nothing [ ConT ''TS.Unmarshal, ConT ''Eq, ConT ''Ord, ConT ''Show, ConT ''Generic, ConT ''Foldable, ConT ''Traversable, ConT ''Functor ] ]`, it resulted in copious errors sharing this sort of message:

```
TreeSitter/Python/AST.hs:8:1: error:
    • Can't make a derived instance of ‘Foldable RelativeImport’:
        Constructor ‘RelativeImport’ must use the type variable only as the last argument of a data type
    • In the data declaration for ‘RelativeImport’
```

#### Reason for behavior

This is due to our use of `Either`s! As @robrix pointed out, If you have `Either (X a) (Y a)` then `a` is used in a non-final position. We have to have a type variable as the last argument of a given datatype to be able to derive instances of `Foldable`, `Traversabl`, and `Functor` due to the shape of the data these classes operate over.

#### Potential workarounds

1. One way around this involves taking advantage of the equivalence between `Either` and `:+:`. `Either` is `* -> * -> *`—it takes two fully-applied types and gives you a fully-applied type. By contrast, the definition of `:+:` in GHC.Generics is `(* -> *) -> (* -> *) -> (* -> *)`—it takes two type constructors and gives you a type constructor so the type has a slightly different shape, but it turns out that `Either (X a) (Y a)` is equivalent to `(X :+: Y) a`. replacing those `Either`s with `:+:`s will have some other consequences. One consequence is that we’d have to replace the `SymbolMatching` & `Unmarshal` instances for `Either` with ones for `:+:` which is weird because we already define `SymbolMatching` for `f :+: g`, i.e. at kind `* -> *`, but this would be instead for `(f :+: g) a`, i.e. at kind `*`)

2. We could also move away from `Either` to some other sum type

**Comment by robrix** [comment]
cf #142 for the latter workaround

**Comment by aymannadeem** [comment]
Fixed by #214
